File: 1330-longest-arithmetic-subsequence-of-given-difference/longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    unordered_map<int,int>dp;

    int longestSubsequence(vector<int>& arr, int diff) {
        
        int ans=1;
        for(int curr:arr){
            dp[curr]=dp[curr-diff]+1;

            ans=max(ans,dp[curr]);
        }
        return ans;
    }
};

File: dynamic-programming/1330-longest-arithmetic-subsequence-of-given-difference/longest-arithmetic-subsequence-of-given-difference.java
class Solution {
    Map<Integer,Integer> dp = new HashMap<>();
    public int longestSubsequence(int[] arr, int diff) {
        int ans=1;

        for(int num:arr){
            dp.put(num,dp.get(num-diff)==null? 1:dp.get(num-diff)+1);
            ans=Math.max(ans,dp.get(num));
        }
        return ans;
    }
}